### PR TITLE
Add a CRD Upgrade Safety preflight check to the `BundleDeployment` controller

### PR DESCRIPTION
**Pull request by everettraven**
**Description**:
- Adds a new `preflights/crdupgradesafety` package containing the logic for a CRD Upgrade Safety preflight check
- Updates the `BundleDeployment` controller to be able to be configured with preflight checks and runs them if they are configured

### CODE REVIEW

File: cmd/core/main.go
Comment (joelanford) [inline cmd/core/main.go:232]: Nit
```suggestion
	aeClient, err := apiextensionsv1client.NewForConfig(cfg)
```

File: internal/preflights/crdupgradesafety/crdupgradesafety.go
Comment (joelanford) [inline internal/preflights/crdupgradesafety/crdupgradesafety.go:72]: Nit:
```suggestion
		if obj.GetObjectKind().GroupVersionKind() != apiextensionsv1.SchemeGroupVersion.WithKind("CustomResourceDefinition") {
			continue
		}
		// outdent the rest of the logic
```

File: internal/preflights/crdupgradesafety/crdupgradesafety.go
Comment (joelanford) [inline internal/preflights/crdupgradesafety/crdupgradesafety.go:99]: Pretty sure this returns `nil` if there are no non-nil errors in the slice.
```suggestion
	return errors.Join(validateErrors...)
```

Comment (everettraven) [reply internal/preflights/crdupgradesafety/crdupgradesafety.go]: updated in https://github.com/operator-framework/rukpak/pull/873/commits/3d4d5b5c90e98dceb74874d3796c0028167c4327

File: internal/preflights/crdupgradesafety/crdupgradesafety.go
Comment (everettraven) [reply internal/preflights/crdupgradesafety/crdupgradesafety.go]: Updated in https://github.com/operator-framework/rukpak/pull/873/commits/3d4d5b5c90e98dceb74874d3796c0028167c4327

File: cmd/core/main.go
Comment (everettraven) [reply cmd/core/main.go:237]: updated in https://github.com/operator-framework/rukpak/pull/873/commits/3d4d5b5c90e98dceb74874d3796c0028167c4327

File: go.mod
Comment (grokspawn) [inline go.mod:3]: Is there a technical reason for this to bump golang also?  I think we'd prefer to keep potentially disruptive dep bumps (for e.g. k8s/controller-runtime/kubebuilder) discrete from feature/bug PRs in general.

Comment (everettraven) [reply go.mod:3]: The version of kapp we are pulling in as a dependency uses go 1.22

File: internal/controllers/bundledeployment/bundledeployment.go
Comment (grokspawn) [inline internal/controllers/bundledeployment/bundledeployment.go:335]: Is this todo going to be resolved before merging this?

Comment (everettraven) [reply internal/controllers/bundledeployment/bundledeployment.go:335]: Ah, I think that should have already been addressed... I'll double check

File: go.mod
Comment (grokspawn) [reply go.mod:3]: Ah.  I see with 1.21++ it [became mandatory](https://go.dev/ref/mod#go-mod-file-go):
```
The go line must be greater than or equal to the go line of all dependencies.
```

Boo.

Comment (everettraven) [reply go.mod:3]: Yeah, there are more preflight checks we could pull in but it resulted in k8s version bumps to v0.30.z and so I refrained from going further than this PR currently does.

It looks like the tilt stuff is failing on this go bump as well...

File: internal/controllers/bundledeployment/bundledeployment.go
Comment (varshaprasad96) [inline internal/controllers/bundledeployment/bundledeployment.go:104]: Nit: Is this required and present by default for now? If so, it would be nice to add a validation for this in `validateConfig` func, just to avoid programmer's error since we are calling `preflight.Install` directly.

File: internal/controllers/bundledeployment/bundledeployment.go
Comment (varshaprasad96) [inline internal/controllers/bundledeployment/bundledeployment.go:171]: Having preflight here looks good, and not blocking this PR at all. I remember this being mentioned somewhere earlier - but just to refresh, can this be directly integrated as a part of helm-operator-plugin's `ActionInterface`: https://github.com/operator-framework/helm-operator-plugins/blob/ae087c419483d73c37155c02f142ab14b4d8c706/pkg/client/actionclient.go#L55?

There may be tweaks required to get the existing release from the cluster and passing it to action client, but this thought popped up because we *may* probably have to expand checks for Uninstall to, and could be cleaner if the helm client can do it all together.

Comment (everettraven) [reply internal/controllers/bundledeployment/bundledeployment.go:171]: I originally had a PR adding it to the actionclient here: https://github.com/operator-framework/helm-operator-plugins/pull/327

There was some discussion in slack where it was decided that it would be a better approach to place the logic directly in operator-controller/rukpak: https://kubernetes.slack.com/archives/C0181L6JYQ2/p1714509603408259

File: internal/controllers/bundledeployment/bundledeployment.go
Comment (everettraven) [reply internal/controllers/bundledeployment/bundledeployment.go:104]: It is not required. If there are no preflight checks defined neither `preflight.Install()` nor `preflight.Upgrade()` should be called.

File: internal/controllers/bundledeployment/bundledeployment.go
Comment (varshaprasad96) [reply internal/controllers/bundledeployment/bundledeployment.go:171]: Ah Sorry! Now I realize why I remembered it! :see_no_evil:

File: pkg/preflights/crdupgradesafety/crdupgradesafety.go
Comment (everettraven) [reply pkg/preflights/crdupgradesafety/crdupgradesafety.go:43]: Two reasons:
- A release of kapp with these change validations has not been cut (we _could_ pin on a commit)
- requires a k8s bump (I am actively queuing these up so we can add the rest of them as part of this PR)

Comment (rashmigottipati) [reply pkg/preflights/crdupgradesafety/crdupgradesafety.go:43]: Ah that makes sense!

Comment (everettraven) [reply pkg/preflights/crdupgradesafety/crdupgradesafety.go:43]: All the ones present in kapp should be added now!

File: go.mod
Comment (everettraven) [reply go.mod]: Rebased on main which already bumped go and k8s deps.

File: api/v1alpha2/bundledeployment_types.go
Comment (joelanford) [inline api/v1alpha2/bundledeployment_types.go:48]: Do we need another reason, or can we just use `InstallFailed` and `UpgradeFailed`, etc?

File: internal/controllers/bundledeployment/bundledeployment.go
Comment (joelanford) [inline internal/controllers/bundledeployment/bundledeployment.go:520]: Seems like it might make sense to return `desiredRelease` even in line 520. Is there a good reason not to?

And looking at 515-518, I wonder if we should do something like this instead:
```suggestion
	relState := stateUnchanged
	if desiredRelease.Manifest != currentRelease.Manifest ||
 		currentRelease.Info.Status == release.StatusFailed ||
 		currentRelease.Info.Status == release.StatusSuperseded {
 		relState = stateNeedsUpgrade
 	}
 	return currentRelease, desiredRelease, relState, nil
```

File: test/e2e/plain_provisioner_test.go
Comment (joelanford) [inline test/e2e/plain_provisioner_test.go:1205]: Is this happening because we're now doing a dry-run install that is failing?

File: api/v1alpha2/bundledeployment_types.go
Comment (everettraven) [reply api/v1alpha2/bundledeployment_types.go:48]: Good question! I think we could get away with just using `InstallFailed` and `UpgradeFailed`.

File: test/e2e/plain_provisioner_test.go
Comment (everettraven) [reply test/e2e/plain_provisioner_test.go:1205]: Correct

File: internal/controllers/bundledeployment/bundledeployment.go
Comment (everettraven) [reply internal/controllers/bundledeployment/bundledeployment.go:520]: >Seems like it might make sense to return desiredRelease even in line 520. Is there a good reason not to?

No particular reason. I was just trying to keep the impact to the logic pretty minimal and didn't think it would be worth returning the desired release in the case where it wouldn't be used, but that is probably better left to the caller to decide than the function. Will tweak this a bit.

File: api/v1alpha2/bundledeployment_types.go
Comment (everettraven) [reply api/v1alpha2/bundledeployment_types.go]: Updated in the latest push

File: internal/controllers/bundledeployment/bundledeployment.go
Comment (everettraven) [reply internal/controllers/bundledeployment/bundledeployment.go]: updated in the latest push

**Review by joelanford — APPROVED** [review]

**Pull request merged**
